src/abi/argument-encoding.md

```diff
@@ -315,3 +319,106 @@ Calling `foo` with `(1u64, "fuel", true)` :
 6675656c00000000 // "fuel" encoded as per the specs
 0000000000000001 // true
 ```
+
+## Version 1
+
+This version was created to replace the older version 0 described above, and follows three philosophical tenets:
+
+- being self-sufficient: it must be possible to completely decode the original data only using the encoded bytes and the original type (there are no references to data outside the encoded bytes);
+- no overhead: only the bare minimum bytes are necessary to do the encoding. No metadata, headers, etc...;
+- no relation with runtime memory layout: no padding, no alignment, etc...
+
+### Primitive Types
+
+Primitive types will be encoded using the exact number of bits they need:
+
+- `u8`: 1 byte;
+- `u16`: 2 bytes;
+- `u32`: 4 bytes;
+- `u64`: 8 bytes;
+- `u256`: 32 bytes;
+- `b256`: 32 bytes;
+
+### Arrays
+
+Arrays are encoded without any padding or alignment, with one item after the other.
+
+- [T; 1] is encoded [encode(T)];
+- [T; 2] is encoded [encode(T), encode(T)]
+
+### Strings
+
+String arrays are encoded just like arrays, without any overhead.
+
+- `str[1]` = 1 byte
+- `str[2]` = 2 bytes
+- `str[n]` = `n` bytes
+
+String slices do contain their length as u64, and the string itself is encoded packed without alignment or padding.
+
+- `"abc"` = `[0, 0, 0, 0, 0, 0, 0, 3, "a", "b", "c"]`
+
+### Slices
+
+`raw_slice`, also being dynamic, contains their length as u64 and is treated as a "slice of bytes". Each byte is encoded as `u8` (1 byte) and is packed without alignment and padding.
+
+For example, a slice of three bytes like `[0u8, 1u8, 2u8]` will be encoded as bytes `[0, 0, 0, 0, 0, 0, 0, 3, 0, 1, 2]`
+
+### Tuple
+
+Tuples are encoded just like arrays, without any overhead like padding and alignment:
+
+- `(A, B, C)` = `[encode(A), encode(B), encode(C)]`
+
+### Structs (v1)
+
+Structs can be encoded in two ways:
+
+- first, with the automatic implementation;
+- second, with the custom implementation.
+
+Auto implementation follows the same rules as tuples. So we can imagine that
+
+```sway
+struct S {
+    a: A,
+    b: B,
+    c: C
+}
+```
+
+is encoded the same way as the tuple `(A, B, C)`.
+
+Custom implementation allows the developer to choose how a struct is encoded.
```

Comment:
There was some conversation around custom implementation in our alignment call, however it seemed that it was still subject to change. Is there any additional documentation on this, or should it even be included at this stage?

Reply:
I would not focus too much on this at this stage given that this is still up for debate. At the moment the SDK has no way to know how the struct was encoded, so we will use the default implementation.